---
title: "Final Project"
author: "Lily Hiltebeitel"
date: "11/24/2021"
output: html_document
---

## Working with TCGA GBM Data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r}
#install packages we want
if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install(c("TCGAbiolinks", "limma", "edgeR", "glmnet", "factoextra", "FactoMineR", "caret","SummarizedExperiment", "gplots", "survival", "survminer", "RColorBrewer", "gProfileR"), force=TRUE)
```
```{r}
# open the packages we need to analysis our data
library("TCGAbiolinks")
library("limma")
library("edgeR")
library("glmnet")
library("factoextra")
library("FactoMineR")
library("caret")
library("SummarizedExperiment")
library("gplots")
library("survival")
library("survminer")
library("RColorBrewer")
library("gProfileR")
library("genefilter")
```
```{r}
# looking at available TCGA projects to make sure we’re in the right space
GDCprojects = getGDCprojects()
head(GDCprojects[c("project_id", "name")])
```
```{r}
# getting data for glioblastoma multiforme from TCGA
TCGAbiolinks:::getProjectSummary("TCGA-GBM")
```
```{r}
# narrowing down to expression data
query_TCGA = GDCquery(
  project = "TCGA-GBM",
  data.category = "Transcriptome Profiling", 
  experimental.strategy = "RNA-Seq",
  workflow.type = "HTSeq - Counts")
```
```{r}
# viewing GBM results and a more user friendly way
gbm_res = getResults(query_TCGA)
colnames(gbm_res) 
```
```{r}
# getting first 6 types of tissue (possibilities are normal solid tissue, primary tumor, recurrent tumor. Our output for GBM is all primary tumor)
head(gbm_res$sample_type) 
```
```{r}
# visualizing in more readable way
# output tells us the amount of primary tumors, recurrent tumors, and normal tissue
# for our GBM data: 156 primary tumor, 13 recurrent tumor, 5 solid tissue normal
summary(factor(gbm_res$sample_type)) 
```
```{r}
# selecting data just for primary solid tumors and solid tissue normal (cancer vs control)
query_TCGA = GDCquery(
  project = "TCGA-GBM",
  data.category = "Transcriptome Profiling", 
  experimental.strategy = "RNA-Seq",
  workflow.type = "HTSeq - Counts",
  sample.type = c("Primary Tumor", "Solid Tissue Normal"))
```
```{r}
# downloading files specified above
GDCdownload(query = query_TCGA)
```
```{r}
# loading RNASeq data into R and double checking object size via dim (output: 56602 161, which is good and expected)
tcga_data = GDCprepare(query_TCGA)
dim(tcga_data)
```
```{r}
# viewing column names of tcga gbm data (both clinical and expression data)
colnames(colData(tcga_data))
```
```{r}
table(tcga_data@colData$vital_status) # to see how many patients are alive still
# output is 29 alive, 124 dead, 2 not reported
```
```{r}
table(tcga_data@colData$definition) # to see counts of primary solid tumor vs normal tissue
# output is 156 primary solid tumor, 5 solid tissue normal
```
```{r}
table(tcga_data@colData$tissue_or_organ_of_origin) # to see tissue/organ origins 
# output is all Brain, NOS
```
```{r}
table(tcga_data@colData$gender) # to see gender breakdown of data
# output is 54 female, 101 male
```
```{r}
table(tcga_data@colData$race) # to see race breakdown of data
# output is 5 asian, 10 black or african american, 1 not reported, 139 white
# notably, the vast majority of this data comes from white patients
```
```{r}
dim(assay(tcga_data)) # view the matrix of gene expression data (again, 56602 161)
```
```{r}
head(rowData(tcga_data)) # this gets the ensembl id/gene id of the first 6 genes
```
```{r}
# saving the above tcga data as an .rds file for future use
saveRDS(object = tcga_data,
        file = "tcga_data.RDS",
        compress = FALSE)
```
```{r}
# code used when reloading this tcga data
tcga_data = readRDS(file = "tcga_data.RDS")
```

## RNASeq Normalizatoin

```{r}
# Now we will normalize the data from its original raw format. This is done by defining a pipeline.
limma_pipeline = function(
  tcga_data,
  condition_variable,
  reference_group=NULL){
  design_factor = colData(tcga_data)[, condition_variable, drop=T]
  group = factor(design_factor)
  if(!is.null(reference_group)){group = relevel(group, ref=reference_group)}
  design = model.matrix(~ group)
  dge = DGEList(counts=assay(tcga_data),
                 samples=colData(tcga_data),
                 genes=as.data.frame(rowData(tcga_data)))
  # filtering by expression
  keep = filterByExpr(dge,design)
  dge = dge[keep,,keep.lib.sizes=FALSE]
  rm(keep)
  # normalization of data (TMM, then followed by voom)
  dge = calcNormFactors(dge)
  v = voom(dge, design, plot=TRUE)
  # fitting the data and the model
  fit = lmFit(v, design)
  fit = eBayes(fit)
  # showing the top genes
  topGenes = topTable(fit, coef=ncol(design), number=100, sort.by="p")
  return(
    list(
      voomObj=v, 
      fit=fit, 
      topGenes=topGenes 
    )
  )
}
```
```{r}
# selecting normal tissue as the reference group
limma_res = limma_pipeline(
  tcga_data=tcga_data,
  condition_variable="definition",
  reference_group="Solid Tissue Normal"
)
```


Figure 1. Mean-variance trend. This figure plots the standard deviation in expression for each gene, represented by a dot in the figure, against a logarithmic scale of the count size. This shows a downward trend which correlates to a biologically similar data set with low variation contributed from technical collection of sequencing reads. The omission of an upwards trend in the beginning shows low expression genes were adequately filtered out of the data TCGA data set.

```{r}
# saving this as an object in case we want it later
saveRDS(object = limma_res,
        file = "limma_res.RDS",
        compress = FALSE)
```
```{r}
# creating a PCA plot to visualize clustering of tumor tissue vs normal tissue, then calling this plot
plot_PCA = function(voomObj, condition_variable){
  group = factor(voomObj$targets[, condition_variable])
  pca = prcomp(t(voomObj$E))
  plot(pca$x[,1:2],col=group, pch=19)
  legend("bottomleft", inset=.01, levels(group), pch=19, col=1:length(levels(group)))
  return(pca)
}
res_pca = plot_PCA(limma_res$voomObj, "definition")
```


Figure 2. RNA expression profiles. The RNA expression profiles of cells shown above in a principal component analysis shows clustering of normal tissue cells away from cancerous tissue samples. The control cell samples all project farther into the first principal component than any cancerous cell samples and are similar to each other in the second principal component. This shows a different RNA expression profile for control tissue cells from cancerous tissue cells. 

## Training and Testing Elastic Net Model for Prediction Use

```{r}
# getting normalized expression values and making it into a factor
d_mat = as.matrix(t(limma_res$voomObj$E))
d_resp = as.factor(limma_res$voomObj$targets$definition)
```
```{r}
# splitting up data into training and testing
# createDataPartition creates boolean vector (true or false value) used to split matrix into 75% training samples and 25% for testing purposes
# random number generator used to keep results consistent between runs
set.seed(42)
train_ids = createDataPartition(d_resp, p=0.75, list=FALSE)
x_train = d_mat[train_ids, ]
x_test  = d_mat[-train_ids, ]
# x_train and y_train are the data used to train model (x is matrix with normalized expression, y is vector with response variable
y_train = d_resp[train_ids]
y_test  = d_resp[-train_ids]
```
```{r}
# Training this Elastic Net Model, which will generate a prediction of genes
# the below code trains the model based on x_train and y_train
res = cv.glmnet(
  x = x_train,
  y = y_train,
  alpha = 0.5,
  family = "binomial"
)
# the result is an object that holds our model’s coefficients
```
```{r}
# using test dataset to test and make prediction
y_pred = predict(res, newx=x_test, type="class", s="lambda.min")
```
```{r}
# to understand how accurate our model is, we compare model prediction against known values
confusion_matrix = table(y_pred, y_test)
print(confusion_matrix)
```
```{r}
# output of this is 1
print(paste0("Sensitivity: ",sensitivity(confusion_matrix)))
```
```{r}
# output of this is 1, which is good!
print(paste0("Specificity: ",specificity(confusion_matrix)))
```
```{r}
# output of this is 1, which is good!
print(paste0("Precision: ",precision(confusion_matrix)))
```
```{r}
# now, let’s use the model to find genes
res_coef = coef(res, s="lambda.min") 
dim(res_coef)
```
```{r}
head(res_coef) 
```
```{r}
# this selects non-zero coefficients, which indicates genes
res_coef = res_coef[res_coef[,1] != 0,]
head(res_coef)
```
```{r}
# removing first coefficient (it’s not necessary)
# get the names (genes)
res_coef = res_coef[-1]
relevant_genes = names(res_coef)
length(relevant_genes)
```
```{r}
# output is a list of 6 genes in Ensembl annotation
head(relevant_genes)
```
```{r}
# Converting names to common gene names. 
head(limma_res$voomObj$genes)
relevant_gene_names = limma_res$voomObj$genes[relevant_genes,"external_gene_name"]
head(relevant_gene_names) # few select genes (with readable names now)
# output: MAPK9, NDC80, PLEKHA1, DLGAP5, E2F8, TOP2A
```

MAPK9. Kinase that targets transcription factors, regulates proliferation, cell death, differentiation.
NDC80. Necessary for proper chromosome segregation.
PLEKHA1. Localized in plasma membrane, suspected to be important for signaling.
DLGAP5. Cell cycle regulation, differentiation regulator, signaling, linked to liver cancer.
E2F8. Transcription factor that regulates G1 to S cell cycle progression, linked to soft palate cancer and retinal cancer.
TOP2A. Topoisomerase, known target for cancer drugs.

Table 1. Predicted DEGs and their functions. While all functions are different, there is a trend of genes associated with proliferation, cell cycle, and communication.

## Making a Heatmap

```{r}
hmcol = colorRampPalette(rev(brewer.pal(9, "RdBu")))(256)
clust = function(x) hclust(x, method="complete")
dist = function(x) as.dist((1-cor(t(x)))/2)
# green boxes represent DEGs
colorLimmaGenes = ifelse(
  # boolean vector for green/white to differentiate our DEGs
  (relevant_genes %in% limma_res$topGenes$ensembl_gene_id),
  "green", # if true, return green
  "white" # if false, return white
)
# parameters are below
gene_heatmap = heatmap.2(
  t(d_mat[,relevant_genes]),
  scale="row",          # scaling gene values
  density.info="none",  # we don’t want the density plot
  trace="none",         # we also don’t want trace lines
  col=hmcol,            # inserting colors that we selected earlier
  labRow=relevant_gene_names, # using common gene names, not ensembl (for readability)
  RowSideColors=colorLimmaGenes,
  labCol=FALSE,         # no column labels
  ColSideColors=as.character(as.numeric(d_resp)), # response class colosr
  dendrogram="both",    # putting dendrograms to show similarities
  hclust = clust,      # define hierarchical clustering
  distfun = dist,       # using correlation coefficient for distance
  cexRow=.6,            
  margins=c(1,5)        
)
```

Code modified from https://www.costalab.org/wp-content/uploads/2020/11/R_class_D3.html#2_TCGA_data. 

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

